8260366: ExtendedSocketOptions <clinit> can deadlock in some circumstances

jaikiran opened the pull request:
Can I please get a review for this change which proposes to fix the issue reported in https://bugs.openjdk.java.net/browse/JDK-8260366?

The issue relates to the concurrent classloading of `sun.net.ext.ExtendedSocketOptions` and `jdk.net.ExtendedSocketOptions` leading to a deadlock. This is because the `sun.net.ext.ExtendedSocketOptions` in its static block does a `Class.forName("jdk.net.ExtendedSocketOptions")`. The `jdk.net.ExtendedSocketOptions` in its own static block calls the `register` method on `sun.net.ext.ExtendedSocketOptions`. If 2 threads concurrently try loading these classes (one loading the `sun.net.ext.ExtendedSocketOptions` and the other loading `jdk.net.ExtendedSocketOptions`), it can so happen that each one ends up holding a classloading lock in the static block of the respective class, while waiting for the other thread to release the lock, thus leading to a deadlock. The stacktrace posted in the linked JBS issue has the necessary details on the deadlock.

The commit here breaks this deadlock by moving out the `Class.forName("jdk.net.ExtendedSocketOptions")` call from the static block of `sun.net.ext.ExtendedSocketOptions` to the `getInstance()` method, thus lazily loading (on first call to `getInstance()`) and registering the `jdk.net.ExtendedSocketOptions`.

Extra attention needs to be given to the `sun.net.ext.ExtendedSocketOptions#register(ExtendedSocketOptions extOptions)` method. Before the change in this PR, when the `sun.net.ext.ExtendedSocketOptions` would successfully complete loading, it was guaranteed that the registered `ExtendedSocketOptions` would either be the one registered from the `jdk.net.ExtendedSocketOptions` or a `NoExtendedSocketOptions`. There wasn't any window of chance for any code (be it in the JDK or in application code) to call the `sun.net.ext.ExtendedSocketOptions#register` to register any different/other implementation/instance of the `ExtendedSocketOptions`. However, with this change in the PR, there is now a window of chance where any code in the JDK (or even application code?) can potentially call the `sun.net.ext.ExtendedSocketOptions#register` before either the `jdk.net.ExtendedSocketOptions` is loaded or the `sun.net.ext.ExtendedSocketOptions#getInstance()` method is called, thus allowing for some other implementation of the `ExtendedSocketOptions` to be registered. However, I'm not sure if it's a practical scenario - although the `sun.net.ext.ExtendedSocketOptions#register` is marked `public`, the comment on that method and the fact that it resides in an internal, not exposed by default class/module, makes me believe that this `register` method isn't supposed to be called by anyone other than the `jdk.net.ExtendedSocketOptions`. If at all this `register` method is allowed to be called from other places, then due to the change in this PR, additional work needs to be probably done in its implementation to allow for the `jdk.net.ExtendedSocketOptions` to be given first priority(?) to be registered first. I'll need input on whether I should worry about this case or if it's fine in its current form in this PR.

This PR also contains a jtreg test which reproduces the issue and verifies the fix.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8260366](https://bugs.openjdk.java.net/browse/JDK-8260366): ExtendedSocketOptions <clinit> can deadlock in some circumstances


### Reviewers
 * [Daniel Fuchs](https://openjdk.java.net/census#dfuchs) (@dfuch - **Reviewer**)


### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2601/head:pull/2601`
`$ git checkout pull/2601`

dfuch commented on test/jdk/sun/net/ext/ExtendedSocketOptionsTest.java:
```diff
@@ -0,0 +1,94 @@
+ * @bug 8260366
+ * @summary Verify that concurrent classloading of sun.net.ext.ExtendedSocketOptions and
+ * jdk.net.ExtendedSocketOptions doesn't lead to a deadlock
+ * @run testng/othervm --add-exports=java.base/sun.net.ext=ALL-UNNAMED ExtendedSocketOptionsTest
```
The --add-exports should not be needed - have you tried simply adding:

```
@modules java.base/sun.net.ext:+open
                  jdk.net
```

before `@run`?  This also has the additional benefit to declare which modules are required to run the test.

dfuch replied:
Since the test should be fast, I'd also advise to stick several identical @run lines (maybe e.g. 5 of them) to increase the probability of the deadlock to happen...

jaikiran replied:
> Since the test should be fast, I'd also advise to stick several identical @run lines (maybe e.g. 5 of them) to increase the probability of the deadlock to happen...

Done. The PR has been updated to run this test multiple times now.

jaikiran replied:
> The --add-exports should not be needed - have you tried simply adding:
> 
> ```
> @modules java.base/sun.net.ext:+open
>                   jdk.net
> ```
> 
> before `@run`? This also has the additional benefit to declare which modules are required to run the test.

I had forgotten about the `@modules` directive. I have now updated the PR use that instead. The only minor difference between what you suggested and my updated PR is that I decided to use `:open` instead of `:+open` for the `sun.net.ext` package, since I don't use the types in that package at compile time, in that test.

dfuch commented on test/jdk/sun/net/ext/ExtendedSocketOptionsTest.java:
```diff
@@ -0,0 +1,94 @@
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
```
Tests do not need - nor should they have - the "Classpath" exception - please see Copyright notice of other tests in the vicinity.

jaikiran replied:
Thank you for catching this. I've fixed it in the latest update of the PR.

dfuch reviewed (changes requested):
Hi Jaikiran, 

Thanks for taking a go at this one.

At a glance - the proposed fix seems reasonable to me. It would be good to have another pair of eyes (and analysis) on this though. Some issues with the test.

jaikiran commented:
Hello Daniel, thank you for the review. I have updated this PR to incorporate your review suggestions for the test.

vyommani commented:
Changes looks OK to me,  any specific reason for removing "final" specifier from "getInstance" & "register" methods ?.

turbanoff commented on src/java.base/share/classes/sun/net/ext/ExtendedSocketOptions.java:
```diff
@@ -167,27 +167,34 @@ protected ExtendedSocketOptions(Set<SocketOption<?>> options) {
 
-    public static final ExtendedSocketOptions getInstance() { return instance; }
+    public static ExtendedSocketOptions getInstance() {
+        if (instance != null) {
```
May be it's worth to avoid reading `volatile` field twice?

jaikiran replied:
Hello @turbanoff, do you mean why read it twice - once here and once inside the `synchronized` block?

turbanoff replied:
Yes. Once here and once inside `synchronized` block.
Reading `volatile` fields cost _something_ on some architectures, so I think we could optimize it a bit.

jaikiran replied:
Hello @turbanoff, the double read is necessary to correctly avoid any race conditions and is a typical strategy used in cases like these. I am not aware of any other alternate more performant strategy, for code like this.

turbanoff replied:
Let me be more clear: I think that it's enough to have only 2 `volatile` field reads _in worst case_. We can use local variable to avoid more than needed reads.
Current code can perform read 4 times _in worst case_: twice outside `synchronized` block and twice inside `synchronized` block.

There are many examples of similar code in the JDK:
https://github.com/openjdk/jdk/blob/master/src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Util.java#L48
https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/lang/StackFrameInfo.java#L125

jaikiran replied:
Ah! of course, I now see what you mean. I misunderstood your previous comment. I'll update this PR shortly.

jaikiran replied:
I've now updated this PR which this suggested change.

jaikiran commented:
> Changes looks OK to me, any specific reason for removing "final" specifier from "getInstance" & "register" methods ?.

Hello Vyom, thank you for the review. Since those two methods are `static`, the `final` was redundant on them and since this patch was already changing those 2 methods, I decided to remove it while I was at it. However, if you and others feel that this patch shouldn't change it, I will introduce it back.

vyommani commented:
> > Changes looks OK to me, any specific reason for removing "final" specifier from "getInstance" & "register" methods ?.
> 
> Hello Vyom, thank you for the review. Since those two methods are `static`, the `final` was redundant on them and since this patch was already changing those 2 methods, I decided to remove it while I was at it. However, if you and others feel that this patch shouldn't change it, I will introduce it back.

I think it's OK for me. What about improving the test little bit,  your test wants to  load both classes at the same time. Please have a look on modified test.


/*
 * Copyright (c) 2021, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */
import org.testng.Assert;
import org.testng.annotations.Test;

import java.util.concurrent.Callable;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

/**
 * @test 
 * @bug 8260366
 * @summary Verify that concurrent classloading of
 * sun.net.ext.ExtendedSocketOptions and jdk.net.ExtendedSocketOptions doesn't
 * lead to a deadlock
 * @modules java.base/sun.net.ext:open
 * @run testng/othervm ExtendedSocketOptionsTest
 * @run testng/othervm ExtendedSocketOptionsTest
 * @run testng/othervm ExtendedSocketOptionsTest
 * @run testng/othervm ExtendedSocketOptionsTest
 * @run testng/othervm ExtendedSocketOptionsTest
 */
public class ExtendedSocketOptionsTest {

    /**
     * Loads {@code jdk.net.ExtendedSocketOptions} and
     * {@code sun.net.ext.ExtendedSocketOptions} concurrently in a thread of
     * their own and expects the classloading of both those classes to
     * succeed.Additionally, after the classloading is successfully done, calls
     * the sun.net.ext.ExtendedSocketOptions#getInstance() and expects it to
     * return a registered ExtendedSocketOptions instance.
     *
     * @throws java.lang.Exception
     */
    @Test
    public void testConcurrentClassLoad() throws Exception {
        CountDownLatch latch = new CountDownLatch(1);
        final Callable<Class<?>> task1 = new Task("jdk.net.ExtendedSocketOptions", latch);
        final Callable<Class<?>> task2 = new Task("sun.net.ext.ExtendedSocketOptions", latch);
        final ExecutorService executor = Executors.newFixedThreadPool(2);
        try {
            final Future<Class<?>>[] results = new Future[2];

            // submit
            results[0] = executor.submit(task1);
            results[1] = executor.submit(task2);
            latch.countDown();

            // wait for completion
            for (Future<Class<?>> result: results) {
                final Class<?> k = result.get();
                System.out.println("Completed loading " + k.getName());
            }
        } finally {
            executor.shutdownNow();
        }
        // check that the sun.net.ext.ExtendedSocketOptions#getInstance() does indeed return
        // the registered instance
        final Class<?> k = Class.forName("sun.net.ext.ExtendedSocketOptions");
        final Object extSocketOptions = k.getDeclaredMethod("getInstance").invoke(null);
        Assert.assertNotNull(extSocketOptions, "sun.net.ext.ExtendedSocketOptions#getInstance()"
                + " unexpectedly returned null");
    }

    private static class Task implements Callable<Class<?>> {

        private final String className;
        private final CountDownLatch latch;

        private Task(final String className, CountDownLatch latch) {
            this.className = className;
            this.latch = latch;
        }

        @Override
        public Class<?> call() {
            System.out.println(Thread.currentThread().getName() + " loading " + this.className);
            try {
                latch.await();
                return Class.forName(this.className);
            } catch (ClassNotFoundException | InterruptedException e) {
                System.err.println("Failed to load " + this.className);
                throw new RuntimeException(e);
            }
        }
    }
}

jaikiran commented:
> What about improving the test little bit, your test wants to load both classes at the same time. Please have a look on modified test.

Hello Vyom, I think that's a good suggestion to use a latch for deciding when to trigger the classloading. I've taken your input and have made some relatively minor change to the way that latch gets used and updated my PR with that change. The latch now waits for both the tasks to reach the point where they are going to do a `Class.forName` on their respectively class names. This should make the test trigger that classloading in separate threads as simultaneously as possible.

vyommani commented:
I think below change will address Andrey's concern 
public static ExtendedSocketOptions getInstance() {
        ExtendedSocketOptions temp = instance;
        if (temp == null) {
            synchronized (ExtendedSocketOptions.class) {
                try {
                    // If the class is present, it will be initialized which
                    // triggers registration of the extended socket options.
                    Class<?> c = Class.forName("jdk.net.ExtendedSocketOptions");
                } catch (ClassNotFoundException e) {
                    // the jdk.net module is not present => no extended socket options
                    instance = new NoExtendedSocketOptions();
                }
                temp = instance;
            }
        }
        return temp;
    }

dfuch commented on src/java.base/share/classes/sun/net/ext/ExtendedSocketOptions.java:
```diff
@@ -167,27 +167,36 @@ protected ExtendedSocketOptions(Set<SocketOption<?>> options) {
+                Class<?> c = Class.forName("jdk.net.ExtendedSocketOptions");
+            } catch (ClassNotFoundException e) {
+                // the jdk.net module is not present => no extended socket options
+                instance = new NoExtendedSocketOptions();
+            }
+        }
+        return instance;
```
I'd suggest changing these two lines as well: 

```
                // the jdk.net module is not present => no extended socket options
                ext = instance = new NoExtendedSocketOptions();
            }
        }
        return ext;
```

jaikiran replied:
Hello Daniel, I had thought about it in my previous commit. But this won't work, since in the normal case where the `ClassNotFoundException` doesn't get thrown, the `instance` is actually set in the `register` method which gets triggered due to the class load on `jdk.net.ExtendedSocketOptions`. As a result, returning the local `ext` variable won't work in that case, unless of course I do `ext = instance` in both the catch block and outside of it, which would, IMO, defeat the purpose of optimization in that last commit.

I decided to "prove" it with some test case and while doing so I just uncovered that my whole patch has just moved the deadlock to a new location - thread T1 calling `sun.net.ext.ExtendedSocketOptions#getInstance()` and thread T2 calling `Class.forName("jdk.net.ExtendedSocketOptions")` ends up in a deadlock. It's clears why that happens.

I am going to take a step back and come back with a different fix for this one. Thank you everyone for these reviews - they definitely helped.

vyommani replied:
Hi Jaikiran,

I tested with my suggested change and i did not see any deadlock at my local Linux environment. I just ran  test in loop and it worked as expected.

Thanks,
Vyom

jaikiran replied:
Hello Vyom,

The trick is to change the test case to have something like this in the "task":

```
diff --git a/test/jdk/sun/net/ext/ExtendedSocketOptionsTest.java b/test/jdk/sun/net/ext/ExtendedSocketOptionsTest.java
index 0702abf5279..26c8a1384a2 100644
--- a/test/jdk/sun/net/ext/ExtendedSocketOptionsTest.java
+++ b/test/jdk/sun/net/ext/ExtendedSocketOptionsTest.java
@@ -96,7 +96,11 @@ public class ExtendedSocketOptionsTest {
                 classLoadingTriggerLatch.countDown();
                 // wait for the other task to let us know it's ready too, to load the class
                 classLoadingTriggerLatch.await();
-                return Class.forName(this.className);
+                final Class<?> c = Class.forName(this.className);
+                // let's call getInstance on sun.net.ext.ExtendedSocketOptions
+                final Class<?> k = Class.forName("sun.net.ext.ExtendedSocketOptions");
+                final Object extSocketOptions = k.getDeclaredMethod("getInstance").invoke(null);
+                return c;
             } catch (Exception e) {
                 System.err.println("Failed to load " + this.className);
                 throw new RuntimeException(e);
```

Essentially, trigger a call to `sun.net.ext.ExtendedSocketOptions#getInstance` and a classload of `jdk.net.ExtendedSocketOptions` simultaneously from different threads. That will end up with:


```

"pool-1-thread-1" #16 prio=5 os_prio=31 cpu=18.25ms elapsed=120.03s tid=0x00007ff9008c8a00 nid=0x6203 waiting for monitor entry  [0x0000700010c8b000]
   java.lang.Thread.State: BLOCKED (on object monitor)
	at sun.net.ext.ExtendedSocketOptions.register(java.base@17-internal/ExtendedSocketOptions.java:197)
	- waiting to lock <0x000000070f9f8e68> (a java.lang.Class for sun.net.ext.ExtendedSocketOptions)
	at jdk.net.ExtendedSocketOptions.<clinit>(jdk.net@17-internal/ExtendedSocketOptions.java:234)
	at java.lang.Class.forName0(java.base@17-internal/Native Method)
	at java.lang.Class.forName(java.base@17-internal/Class.java:375)
	at ExtendedSocketOptionsTest$Task.call(ExtendedSocketOptionsTest.java:100)
	at ExtendedSocketOptionsTest$Task.call(ExtendedSocketOptionsTest.java:84)
	at java.util.concurrent.FutureTask.run(java.base@17-internal/FutureTask.java:264)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(java.base@17-internal/ThreadPoolExecutor.java:1135)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(java.base@17-internal/ThreadPoolExecutor.java:635)
	at java.lang.Thread.run(java.base@17-internal/Thread.java:831)

"pool-1-thread-2" #17 prio=5 os_prio=31 cpu=8.76ms elapsed=120.03s tid=0x00007ff90102be00 nid=0x6403 in Object.wait()  [0x0000700010d8e000]
   java.lang.Thread.State: RUNNABLE
	at java.lang.Class.forName0(java.base@17-internal/Native Method)
	- waiting on the Class initialization monitor for jdk.net.ExtendedSocketOptions
	at java.lang.Class.forName(java.base@17-internal/Class.java:375)
	at sun.net.ext.ExtendedSocketOptions.getInstance(java.base@17-internal/ExtendedSocketOptions.java:185)
	- locked <0x000000070f9f8e68> (a java.lang.Class for sun.net.ext.ExtendedSocketOptions)
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(java.base@17-internal/Native Method)
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(java.base@17-internal/NativeMethodAccessorImpl.java:78)
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(java.base@17-internal/DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(java.base@17-internal/Method.java:566)
	at ExtendedSocketOptionsTest$Task.call(ExtendedSocketOptionsTest.java:102)
	at ExtendedSocketOptionsTest$Task.call(ExtendedSocketOptionsTest.java:84)
	at java.util.concurrent.FutureTask.run(java.base@17-internal/FutureTask.java:264)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(java.base@17-internal/ThreadPoolExecutor.java:1135)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(java.base@17-internal/ThreadPoolExecutor.java:635)
	at java.lang.Thread.run(java.base@17-internal/Thread.java:831)

```

dfuch replied:
> Hello Daniel, I had thought about it in my previous commit. But this won't work, since in the normal case where the `ClassNotFoundException` doesn't get thrown, the `instance` is actually set in the `register` method which gets triggered due to the class load on `jdk.net.ExtendedSocketOptions`. As a result, returning the local `ext` variable won't work in that case, unless of course I do `ext = instance` in both the catch block and outside of it, which would, IMO, defeat the purpose of optimization in that last commit.

Oh - right - still - if you'd stick with that fix - I believe you should do that:

```
            try {
                // If the class is present, it will be initialized which
                // triggers registration of the extended socket options.
                Class<?> c = Class.forName("jdk.net.ExtendedSocketOptions");
                ext = instance;
            } catch (ClassNotFoundException e) {
                // the jdk.net module is not present => no extended socket options
                ext = instance = new NoExtendedSocketOptions();
            }
        }
        return ext;
 ```
 
 Anyway: waiting for the next fix :-)

jaikiran commented:
I'm closing this for now, for the reason stated here https://github.com/openjdk/jdk/pull/2601#discussion_r578400669.

jaikiran commented:
After thinking a bit more about this issue and the patch I had proposed, I realized what I did wrong in this patch. The `synchronized` block in the `sun.net.ext.ExtendedSocketOptions#getInstance()` was scoped one statement too many. We don't (and shouldn't) need the `Class.forName("jdk.net.ExtendedSocketOptions")` to be part of that synchronized block. I've now modified and fixed that part to have the synchronized block only at the place where it should be needed. I've also enhanced the test to introduce a couple of additional concurrent tasks which now call `sun.net.ext.ExtendedSocketOptions#getInstance()`. So in total, the test now fires off 4 tasks, 2 of which load the respective classes and the other 2 call the getInstance() method, all concurrently. This should make the test a bit more robust and should catch any potential deadlocks. With this fix, the test now passes.

I've now updated and reopened this PR for further reviews.

vyommani commented:
> After thinking a bit more about this issue and the patch I had proposed, I realized what I did wrong in this patch. The `synchronized` block in the `sun.net.ext.ExtendedSocketOptions#getInstance()` was scoped one statement too many. We don't (and shouldn't) need the `Class.forName("jdk.net.ExtendedSocketOptions")` to be part of that synchronized block. I've now modified and fixed that part to have the synchronized block only at the place where it should be needed. I've also enhanced the test to introduce a couple of additional concurrent tasks which now call `sun.net.ext.ExtendedSocketOptions#getInstance()`. So in total, the test now fires off 4 tasks, 2 of which load the respective classes and the other 2 call the getInstance() method, all concurrently. This should make the test a bit more robust and should catch any potential deadlocks. With this fix, the test now passes.
> 
> I've now updated and reopened this PR for further reviews.

Hi Jaikiran,

I looked into your previous patch and i believe why you were observing the deadlock because you make "sun.net.ext.ExtendedSocketOption.register" thread safe(synchronize). When you are loading both the classes simultaneously one thread is calling getInstance()  and other thread while loading the class(jdk.net.ExtendedSocketOptions) calls  register  and you are observing the deadlock.

Although sun.net.ext.ExtendedSocketOptions.register is public method but it will be called only from jdk.net.ExtendedSocketOption static block, so we don't need it to be thread safe.

Let's wait for what others people say.

jaikiran commented:
Any reviews/comments, in addition to what Vyom mentioned, to the updated PR please?

dfuch reviewed (approved):
The fix and test looks good to me.

jaikiran commented:
Thank you for the reviews, Daniel.

jaikiran commented:
/integrate
